src/client/desktop_gui/lib/vms_table.dart

```diff
@@ -0,0 +1,375 @@
+import 'dart:convert';
+
+import 'package:basics/iterable_basics.dart';
+import 'package:collection/collection.dart';
+import 'package:data_table_2/data_table_2.dart';
+import 'package:filesize/filesize.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+import 'grpc_client.dart';
+import 'instance_actions_button.dart';
+import 'settings.dart';
+import 'vms_screen.dart';
+
+const columnWidthsKey = 'columnWidths';
+
+class TableHeader {
+  final String title;
+  double width;
+  final double minWidth;
+  final String Function(VmInfo)? sortExtractor;
+  void Function(double)? resizeCallback;
+  void Function()? resizeStartCallback;
+
+  TableHeader({
+    required this.title,
+    required this.width,
+    required this.minWidth,
+    this.sortExtractor,
+    this.resizeCallback,
+    this.resizeStartCallback,
+  });
+}
+
+class VMsTable extends ConsumerStatefulWidget {
+  const VMsTable({super.key});
+
+  @override
+  ConsumerState<VMsTable> createState() => _VMsTableState();
+}
+
+class _VMsTableState extends ConsumerState<VMsTable> {
+  int? sortIndex;
+  bool sortAscending = true;
+  String Function(VmInfo)? sortExtractor;
+
+  double tableWidthFactor = 0;
+
+  // the widths represent fractions out of the total sum
+  // the minWidths are the minimum widths but in pixels!
+  final headers = [
+    TableHeader(
+      title: 'Name',
+      width: 115,
+      minWidth: 60,
+      sortExtractor: (VmInfo info) => info.name,
+    ),
+    TableHeader(
+      title: 'State',
+      width: 100,
+      minWidth: 65,
+      sortExtractor: (VmInfo info) => info.instanceStatus.status.name,
+    ),
+    TableHeader(title: 'Memory usage', width: 140, minWidth: 115),
+    TableHeader(title: 'Disk usage', width: 130, minWidth: 90),
+    TableHeader(
+      title: 'CPU(s)',
+      width: 80,
+      minWidth: 75,
+      sortExtractor: (VmInfo info) => info.cpuCount,
+    ),
+    TableHeader(title: 'Load', width: 100, minWidth: 50),
+    TableHeader(
+      title: 'Image',
+      width: 125,
+      minWidth: 70,
+      sortExtractor: (VmInfo info) => info.currentRelease,
+    ),
+    TableHeader(
+      title: 'Ipv4',
+      width: 143,
+      minWidth: 80,
+      sortExtractor: (VmInfo info) => info.ipv4.firstOrNull ?? '',
+    ),
+    TableHeader(
+      title: 'Actions',
+      width: 67,
+      minWidth: 60,
+    ),
+  ];
+
+  double get headersWidth => headers.map((h) => h.width).sum;
+
+  DataColumnSortCallback _extractorToSortCallback(
+    String Function(VmInfo) extractor,
+  ) =>
+      (index, ascending) => setState(() {
+            sortIndex = (sortIndex == index && ascending) ? null : index;
+            sortAscending = ascending;
+            sortExtractor = extractor;
+          });
+
+  Widget _textWithEllipsis(String text) {
+    return Tooltip(
+      message: text,
+      child: Text(
+        text.replaceAll('-', '\u2011').replaceAll(' ', '\u00A0'),
```

Comment:
It has to do with text overflow. Flutter **_really_** want to break the text at word boundaries in case of overflow (we can't tell it not to do that). But in our case, I don't think we want that. For example, if an instance name is 'short-looooooooooong', it would get clipped as 'short-...', even though we could fit 'short-looooo...', and I think the latter is more useful. These string replacements add non-breaking space and non-breaking hyphen, so that Flutter considers them part of the word, not a word boundary, and so it clips only what doesn't fit.